### PR TITLE
Add redis for caching and rate limiting

### PR DESCRIPTION
**Pull request by lossyrob**
## Description

This PR adds redis as cache storage. It also uses the lua script and logic from https://github.com/long2ice/fastapi-limiter to enable rate limiting on STAC API endpoints.

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Tested locally and on a deployed dev stack.

## Checklist:

Please delete options that are not relevant.

- [x] I have performed a self-review
- [x] Changelog has been updated
- [x] Unit tests pass locally (./scripts/test)
- [x] Code is linted and styled (./scripts/format)

### CODE REVIEW

File: pccommon/setup.py [hunk 1]
```diff
@@ -14,6 +14,7 @@
     "cachetools==5.0.0",
     "types-cachetools==4.2.9",
     "pyhumps==3.5.3",
+    "redis==4.2.0-rc1",
```
Comment (lossyrob) [inline pccommon/setup.py:17]: 4.2.0-rc1 is the first version with async capabilities. Should move to released version when available.

File: pcstac/pcstac/client.py [hunk 2]
```diff
@@ -84,10 +92,14 @@ async def all_collections(self, **kwargs: Dict[str, Any]) -> Collections:
         Returns:
             Collections.
         """
-        try:
-            return collections_endpoint_cache["/collections"]
-        except KeyError:
-            collections = await super().all_collections(**kwargs)
+        await rate_limit(
```
Comment (mmcfarland) [inline pcstac/pcstac/client.py:95]: Is it possible to convert this to a decorator that could be applied to these functions? Could be nice from a portability/visibility perspective.

**Review by mmcfarland — APPROVED** [review]
Looks great - tested things out locally but couldn't assess any speed improvements for cached results (local -> staging.db), but functionality was consistent. I was able to generate 429s with a simple `ab` command, so that part looks solid.

File: pcstac/pcstac/client.py (hunk 2 above)
Comment (lossyrob) [reply pcstac/pcstac/client.py]: Good call - I did this for rate_limit and also the new back_pressure.

**Review by mmcfarland — COMMENTED** [review]
Changes look good to me. I wasn't able to construct a test case where I could get overage to be > 0 without hitting 429s, but I think that's just a limitation of the tool I was using (apache bench). Would be interesting in invest some time into some scriptable load testing using something like [k6](https://k6.io/docs/testing-guides/api-load-testing/) at some point.

Are you planning on addressing Hub-cluster request exceptions in this PR, or is that follow up work?

**Comment by lossyrob** [comment]
Planning on investigating the Hub IP issues before merging.

I was able to test tings locally with the skipped test, and with different configurations of the timeout for the items endpoint to cause the slowdown to trigger, and eventually be overloaded to the rate limit. Also tested the dev deployment and locally using [vegeta](https://github.com/tsenart/vegeta) which showed eventual 429's. Agree about the scriptable load testing - let's chat about implementing this for testing these changes once they reach staging.

**Review by mmcfarland — APPROVED** [review]
Looks good. I tested by adding my local ip and was able to hit a different threshold before returning 429s. Note I had to add https://github.com/microsoft/planetary-computer-apis/pull/52/commits/6134d4fbb10162ead3e73792174575598b2af529 to get the local service started (may have missed those lines in your commit?)

**Pull request merged**